### ISSUE TITLE
The algorithm does not find trends in the final intervals

### ISSUE THREAD

**Issue by jackure** [alvarobartt/trendet#11]
What makes an algorithm like this attractive is being able to identify the current trend of the market.
However, in all tests I've done, the algorithm ignores the final intervals.
Below I show you the result of the algorithm with smoothed data, to further accentuate the phenomenon that I mentioned.

![image](https://user-images.githubusercontent.com/3168913/82735295-d4a02200-9d20-11ea-9c40-b155654b79e1.png)

**Comment by alvarobartt** [comment]
Hi @jackure, thank you for reporting this!

I already need to perform some tests over this so as to identify the problem and find the best possible solution! Please, could you send me a CSV file containing the data that produced that concrete error? Or just the trendet function call that you made?

__Thank you all again! Remember to star the repo if you found it useful and follow me at GitHub so as to get notified of all the updates!__

**Comment by mattstone** [comment]
Just adding to this.. something does look a bit broken with the trend identification algorithm.

This is the test data I've used.

test = investpy.get_stock_historical_data(stock='AAPL',
                                          country='united states',
                                          from_date='01/04/2020',
                                          to_date='20/07/2020')


df = trendet.identify_df_trends(df=test, column='Close')

This is the result.

<img width="981" alt="Screen Shot 2020-07-26 at 10 15 56 am" src="https://user-images.githubusercontent.com/20787/88468589-3ea79400-cf29-11ea-8063-06538f78f2e3.png">


Also, the demonstration breaks if there is no downtrend (as in the above example). Quick fix highlighted in bold.

with plt.style.context('seaborn-paper'):
    plt.figure(figsize=(20, 10))

    ax = sns.lineplot(x=df['Date'], y=df['Close'])

    **if 'Up Trend' in df.columns:**
      labels = df['Up Trend'].dropna().unique().tolist()

      for label in labels:
          sns.lineplot(x=df[df['Up Trend'] == label]['Date'],
                       y=df[df['Up Trend'] == label]['Close'],
                       color='green')

          ax.axvspan(df[df['Up Trend'] == label]['Date'].iloc[0],
                     df[df['Up Trend'] == label]['Date'].iloc[-1],
                     alpha=0.2,
                     color='green')

    **if 'Down Trend' in df.columns:**
      labels = df['Down Trend'].dropna().unique().tolist()

      for label in labels:
          sns.lineplot(x=df[df['Down Trend'] == label]['Date'],
                       y=df[df['Down Trend'] == label]['Close'],
                       color='red')

          ax.axvspan(df[df['Down Trend'] == label]['Date'].iloc[0],
                     df[df['Down Trend'] == label]['Date'].iloc[-1],
                     alpha=0.2,
                     color='red')

    plt.show()

**Comment by sunilhariharan** [comment]
Hi…i have fixed this….can i post an image of my code /the modified code ?

**Comment by yuvalfoox** [comment]
@sunilhariharan Can you please share your code?

**Comment by sunilhariharan** [comment]
![Uploading B78C2FB1-1E3E-482A-B58A-B8F79D979A52.jpeg…]()

**Comment by AliAlhajji** [comment]
I noticed the same issue too. When I change the ```to_date``` parameter to be in the middle of an area that was predicted as a trend, it does not identify it as a  trend anymore.

Example:

![image](https://user-images.githubusercontent.com/35175106/128876894-e75b4a46-2fb3-40cf-b45f-2d987c6c43d4.png)


And when I changed ```to_date``` to be in 30/06/2021, the uptrend was not identified:
![image](https://user-images.githubusercontent.com/35175106/128877401-49da5662-c9e2-4107-ba54-4440749d9604.png)


You can try it with this stock:

Stock: 8240
Country: Saudi Arabia
From: "01/02/2021"
To (1):  "08/08/2021"
To (2): "30/06/2021"
Window: 5

**Comment by sunilhariharan** [comment]
The bug is in the for loop…This algo will never be able to find the last upward and the last downward trend.
In the for loop,when the limit is being compared to the value, what is happening is for the last updard n downward trend iteration, the algorithm is not able to get a value

**Comment by sunilhariharan** [comment]
The bug is in the for loop…This algo will never be able to find the last upward and the last downward local trend.
In the for loop,when the limit is being compared to the value, what is happening is for the last updard n downward trend iteration, the algorithm is not able to get a value greater than the limit (in the elif part)…so it is not appending that trend

**Comment by AliAlhajji** [comment]
> The bug is in the for loop…This algo will never be able to find the last upward and the last downward local trend.
> In the for loop,when the limit is being compared to the value, what is happening is for the last updard n downward trend iteration, the algorithm is not able to get a value greater than the limit (in the elif part)…so it is not appending that trend

Were you able to fix it? Can you share the code if you did?

Thanks.

**Comment by sunilhariharan** [comment]
![WhatsApp Image 2021-08-11 at 12 01 21 AM (6)](https://user-images.githubusercontent.com/42323117/128916102-440498c2-d049-410e-8456-5caa8e73b04f.jpeg)

**Comment by sunilhariharan** [comment]
![WhatsApp Image 2021-08-11 at 12 01 21 AM](https://user-images.githubusercontent.com/42323117/128916174-7fbe71fe-a24a-4bdc-a897-0b088beb9dc0.jpeg)
![WhatsApp Image 2021-08-11 at 12 01 21 AM (1)](https://user-images.githubusercontent.com/42323117/128916185-478f24aa-66f7-4610-aa71-b819053dde94.jpeg)
![WhatsApp Image 2021-08-11 at 12 01 21 AM (2)](https://user-images.githubusercontent.com/42323117/128916196-9153364f-2faa-4cbf-aa53-e1f5dcff5b89.jpeg)
![WhatsApp Image 2021-08-11 at 12 01 21 AM (3)](https://user-images.githubusercontent.com/42323117/128916204-987ccc64-969f-4c55-8330-d78757f97178.jpeg)
![WhatsApp Image 2021-08-11 at 12 01 21 AM (4)](https://user-images.githubusercontent.com/42323117/128916215-13c20670-d1bc-4cf4-b98d-a84287430f9c.jpeg)
![WhatsApp Image 2021-08-11 at 12 01 21 AM (5)](https://user-images.githubusercontent.com/42323117/128916228-dc6e2c73-f31a-41ba-82ea-f5b38322bfaa.jpeg)

**Comment by AliAlhajji** [comment]
Thansk @sunilhariharan !
If the code is working as expected, would you like to share it as a forked repo or a pull request?
It's hard to follow line by line here :)

**Comment by AliAlhajji** [comment]
Or at least copy and paste the code in here so that we can follow your logic easily.

@sunilhariharan

**Comment by yuvalfoox** [comment]
@sunilhariharan 
Thanks for sharing!
Can you please share it as a repo or copy the full text as comment with example data?
I got a bit lost there :|
Thanks!

**Comment by jokaorgua** [comment]
@sunilhariharan could you share the code in text please?

**Comment by jokaorgua** [comment]
@alvarobartt I've made a pull request. please check it and if everything is ok - add to master

**Comment by AliAlhajji** [comment]
> @alvarobartt I've made a pull request. please check it and if everything is ok - add to master

I tried your code. I don't think it's doing better than what we have now. I've back-tested it on several stocks, none of which showed uptrend towards the end even if there was one.

**Comment by jokaorgua** [comment]
> > @alvarobartt I've made a pull request. please check it and if everything is ok - add to master
> 
> I tried your code. I don't think it's doing better than what we have now. I've back-tested it on several stocks, none of which showed uptrend towards the end even if there was one.

Could you provide example data so I can backtest it too ?

**Comment by sunilhariharan** [comment]
@AliAlhajji did you try my code?

**Comment by erdemustun** [comment]
@sunilhariharan 

Is it possible for you to share the entire code you added to the Trendet app?

Data can no longer be retrieved from Investing. It would be more convenient to provide it through Yahoo Finance.

--> 653 df_ans,v1,results,trends,from_trend,objs=identify_df_trends(df_fin,column='trend')
    654 
    655 values=v1[-1]

NameError: name 'df_fin' is not defined

**Comment by erdemustun** [comment]
@alvarobartt 
Your trendet code is very well thought out and written, but it does not find a trend in the last part of the data.

<img width="1481" height="687" alt="Image" src="https://github.com/user-attachments/assets/604f3c35-998e-47fe-b46a-17a5e7e1458e" />

start_date = '2024-08-01'
end_date = '2025-08-30'
tickers=['AAPL']
